Generic traits can't use `where` clause for trait bounds

# Problem

It appears that if you are trying to use a generic trait, the generic trait cannot have it's trait bounds defined in a `where` clause. Ambassador **does** still work when using the short-form angle bracket notation though.

Is this an already known issue? It seems like something that would have been listed as a known limitation or have an issue open for it already but I couldn't find anything.

Side Note: The code below is a "Minimal Reproducible Example" and it took me quite a while to figure out why my actual code didn't work but the simple example I made (happening to not use the `where` clause in the trait) did 😅

# Information

## Failing Code

```rust
use ambassador::{Delegate, delegatable_trait};

trait GenericType {}

#[delegatable_trait]
trait GenericJob<T>
where
    T: GenericType,
{
}

struct ConcreteType;
impl GenericType for ConcreteType {}

struct Job1;
impl GenericJob<ConcreteType> for Job1 {}

#[derive(Delegate)]
#[delegate(GenericJob<T>, generics = "T", where = "T: GenericType")]
enum ConcreteJob {
    Job1(Job1),
}

fn main() {
    ConcreteJob::Job1(Job1 {});
}


```

## Error
```
error[E0277]: the trait bound `T: GenericType` is not satisfied
  --> src/main.rs:6:7
   |
6  | trait GenericJob<T>
   |       ^^^^^^^^^^ the trait `GenericType` is not implemented for `T`
...
18 | #[derive(Delegate)]
   |          -------- in this derive macro expansion
   |
note: required by a bound in `GenericJob`
  --> src/main.rs:8:8
   |
6  | trait GenericJob<T>
   |       ---------- required by a bound in this trait
7  | where
8  |     T: GenericType,
   |        ^^^^^^^^^^^ required by this bound in `GenericJob`
   = note: this error originates in the macro `ambassador_impl_GenericJob` which comes from the expansion of the derive macro `Delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `T` with trait `GenericType`
   |
6  | trait GenericJob<T: GenericType>
   |                   +++++++++++++
```
## Expanded Failing Code

``` rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2024::*;
#[macro_use]
extern crate std;
use ambassador::{Delegate, delegatable_trait};
trait GenericType {}
trait GenericJob<T>
where
    T: GenericType,
{}
#[doc(inline)]
///A macro to be used by [`ambassador::Delegate`] to delegate [`GenericJob`]
use _ambassador_impl_GenericJob as ambassador_impl_GenericJob;
#[doc(hidden)]
#[allow(non_snake_case)]
mod ambassador_impl_GenericJob {}
struct ConcreteType;
impl GenericType for ConcreteType {}
struct Job1;
impl GenericJob<ConcreteType> for Job1 {}
#[delegate(GenericJob<T>, generics = "T", where = "T: GenericType")]
enum ConcreteJob {
    Job1(Job1),
}
#[allow(non_snake_case)]
mod ambassador_module_GenericJob_for_ConcreteJob {
    use super::*;
    #[doc(hidden)]
    #[allow(non_camel_case_types)]
    pub trait MatchGenericJob<T, ambassador_X: GenericJob<T>>: GenericJob<T> {}
    #[allow(non_camel_case_types)]
    impl<
        T,
        ambassador_X: GenericJob<T>,
        ambassador_Y: GenericJob<T>,
    > MatchGenericJob<T, ambassador_X> for ambassador_Y {}
    impl<T> GenericJob<T> for ConcreteJob
    where
        T: GenericType,
        Job1: GenericJob<T>,
    {}
}
fn main() {
    ConcreteJob::Job1(Job1 {});
}
```

---

## Working Code

```rust
use ambassador::{Delegate, delegatable_trait};

trait GenericType {}

#[delegatable_trait]
trait GenericJob<T: GenericType> {}

struct ConcreteType;
impl GenericType for ConcreteType {}

struct Job1;
impl GenericJob<ConcreteType> for Job1 {}

#[derive(Delegate)]
#[delegate(GenericJob<T>, generics = "T", where = "T: GenericType")]
enum ConcreteJob {
    Job1(Job1),
}

fn main() {
    ConcreteJob::Job1(Job1 {});
}
```

## Expanded Working Code

```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2024::*;
#[macro_use]
extern crate std;
use ambassador::{Delegate, delegatable_trait};
trait GenericType {}
trait GenericJob<T: GenericType> {}
#[doc(inline)]
///A macro to be used by [`ambassador::Delegate`] to delegate [`GenericJob`]
use _ambassador_impl_GenericJob as ambassador_impl_GenericJob;
#[doc(hidden)]
#[allow(non_snake_case)]
mod ambassador_impl_GenericJob {}
struct ConcreteType;
impl GenericType for ConcreteType {}
struct Job1;
impl GenericJob<ConcreteType> for Job1 {}
#[delegate(GenericJob<T>, generics = "T", where = "T: GenericType")]
enum ConcreteJob {
    Job1(Job1),
}
#[allow(non_snake_case)]
mod ambassador_module_GenericJob_for_ConcreteJob {
    use super::*;
    #[doc(hidden)]
    #[allow(non_camel_case_types)]
    pub trait MatchGenericJob<
        T: GenericType,
        ambassador_X: GenericJob<T>,
    >: GenericJob<T> {}
    #[allow(non_camel_case_types)]
    impl<
        T: GenericType,
        ambassador_X: GenericJob<T>,
        ambassador_Y: GenericJob<T>,
    > MatchGenericJob<T, ambassador_X> for ambassador_Y {}
    impl<T> GenericJob<T> for ConcreteJob
    where
        T: GenericType,
        Job1: GenericJob<T>,
    {}
}
fn main() {
    ConcreteJob::Job1(Job1 {});
}
```

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -22,7 +22,7 @@ jobs:
           toolchain: ${{ matrix.rust }}
           override: true
       - run: rustup component add rustfmt
-      - run: cargo +nightly update -Z minimal-versions
+      - run: cargo +nightly-2021-06-17 update -Z minimal-versions
       - uses: actions-rs/cargo@v1
         with:
           command: fmt
@@ -48,7 +48,7 @@ jobs:
           toolchain: ${{ matrix.rust }}
           override: true
       - run: rustup component add clippy
-      - run: cargo +nightly update -Z minimal-versions
+      - run: cargo +nightly-2021-06-17  update -Z minimal-versions
       - uses: actions-rs/cargo@v1
         with:
           command: clippy

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -22,7 +22,7 @@ jobs:
           profile: minimal
           toolchain: ${{ matrix.rust }}
           override: true
-      - run: cargo +nightly update -Z minimal-versions
+      - run: cargo +nightly-2021-06-17 update -Z minimal-versions
       - uses: actions-rs/cargo@v1
         with:
           command: check

--- a/ambassador/src/register.rs
+++ b/ambassador/src/register.rs
@@ -147,6 +147,7 @@ fn make_assoc_ty_bound(
 ) -> TokenStream {
     let trait_ident = &item_trait.ident;
     let gen_params = &item_trait.generics.params;
+    let gen_where = &item_trait.generics.where_clause;
     let gen_params_t = super::util::TailingPunctuated(gen_params);
 
     let gen_tokens: TokenStream = gen_params.iter().flat_map(param_to_tokens).collect();
@@ -166,9 +167,9 @@ fn make_assoc_ty_bound(
     quote! {
         #[doc(hidden)]
         #[allow(non_camel_case_types)]
-        pub trait #match_name<#gen_params_t ambassador_X: #trait_ident<#gen_tokens>>: #new_bound {}
+        pub trait #match_name<#gen_params_t ambassador_X: #trait_ident<#gen_tokens>>: #new_bound #gen_where {}
         #[allow(non_camel_case_types)]
-        impl<#gen_params_t ambassador_X: #trait_ident<#gen_tokens>, ambassador_Y: #new_bound> #match_name<#gen_tokens ambassador_X> for ambassador_Y {} // Replace with trait alias when they become stable
+        impl<#gen_params_t ambassador_X: #trait_ident<#gen_tokens>, ambassador_Y: #new_bound> #match_name<#gen_tokens ambassador_X> for ambassador_Y #gen_where {}
     }
 }
 

--- a/ambassador/tests/run-pass/trait_where.rs
+++ b/ambassador/tests/run-pass/trait_where.rs
@@ -1,0 +1,27 @@
+extern crate ambassador;
+use ambassador::{delegatable_trait, Delegate};
+
+trait GenericType {}
+
+#[delegatable_trait]
+trait GenericJob<T>
+where
+    T: GenericType,
+{
+}
+
+struct ConcreteType;
+impl GenericType for ConcreteType {}
+
+struct Job1;
+impl GenericJob<ConcreteType> for Job1 {}
+
+#[derive(Delegate)]
+#[delegate(GenericJob<T>, generics = "T", where = "T: GenericType")]
+enum ConcreteJob {
+    Job1(Job1),
+}
+
+fn main() {
+    ConcreteJob::Job1(Job1 {});
+}

